### DISCUSSION TITLE
TryGetValue_returns_false_if_source_is_empty_and_out_parameter_is_null

### DISCUSSION THREAD

**Discussion by xavierjohn** [vkhorikov/CSharpFunctionalExtensions#482] [Q&A] [answered]
Is this useful in real scenarios? 
TryGetValue_returns_false_if_source_is_empty_and_out_parameter_is_null

https://github.com/vkhorikov/CSharpFunctionalExtensions/blob/7540e4ff91a64a4a9bbfc35ff18cf12d86c98c0b/CSharpFunctionalExtensions.Tests/MaybeTests/BasicTests.cs#L200

Why would we allow nullable type inside a MayBe?
Maybe<int?>

The purpose of Maybe is to say Value or nothing. 

Value = int, HasValue = true
OR
Value = _, HasValue = false

I am trying to imagine where I would want to do something like
```csharp
var maybe = Maybe<int?>.None;
bool result = maybe.TryGetValue(out int? value);
```
vs

```csharp
var maybe = Maybe<int>.None;
bool result = maybe.TryGetValue(out int value);
```

**Comment by vkhorikov** [comment, marked as answer]
This is probably not a useful scenario in general, but still, there could be usecases where you are using `Maybe<T>` without being able to control the type of `T` (which may turn into `int?`).

Also, I don't think there's a way to restrict the use a generic type to non-nullable types only. `?` is a shortcut for `Nullable<T>` which itself is a struct.

**Comment by xavierjohn** [comment]
I found having T? _value within Maybe where T is int? confusing so my friend gave an implementation of maybe that will not allows nulls. For now I will keep the restriction till some real world scenario needs nullable int. 

```csharp
public readonly struct Maybe<T> :
    IEquatable<T>,
    IEquatable<Maybe<T>>
    where T : notnull
{
```
Full implementation here.
https://github.com/xavierjohn/FunctionalDDD/blob/main/RailwayOrientedProgramming/src/Maybe/Maybe%7BT%7D.cs

I am also not sure what value the interface IMaybe adds.
